// @aws-cdk/assertions/lib/private/resources.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.countResources = exports.hasResourceProperties = exports.hasResource = exports.findResources = void 0;
const __1 = require("..");
const absent_1 = require("./matchers/absent");
const section_1 = require("./section");
function findResources(template, type, props = {}) {
    var _a;
    const section = (_a = template.Resources) !== null && _a !== void 0 ? _a : {};
    const result = section_1.matchSection(filterType(section, type), props);
    if (!result.match) {
        return {};
    }
    return result.matches;
}
exports.findResources = findResources;
function hasResource(template, type, props) {
    var _a;
    const section = (_a = template.Resources) !== null && _a !== void 0 ? _a : {};
    const result = section_1.matchSection(filterType(section, type), props);
    if (result.match) {
        return;
    }
    if (result.closestResult === undefined) {
        return `No resource with type ${type} found`;
    }
    return [
        `Template has ${result.analyzedCount} resources with type ${type}, but none match as expected.`,
        section_1.formatFailure(result.closestResult),
    ].join('\n');
}
exports.hasResource = hasResource;
function hasResourceProperties(template, type, props) {
    // amended needs to be a deep copy to avoid modifying the template.
    let amended = JSON.parse(JSON.stringify(template));
    // special case to exclude AbsentMatch because adding an empty Properties object will affect its evaluation.
    if (!__1.Matcher.isMatcher(props) || !(props instanceof absent_1.AbsentMatch)) {
        amended = addEmptyProperties(amended);
    }
    return hasResource(amended, type, __1.Match.objectLike({
        Properties: props,
    }));
}
exports.hasResourceProperties = hasResourceProperties;
function countResources(template, type) {
    var _a;
    const section = (_a = template.Resources) !== null && _a !== void 0 ? _a : {};
    const types = filterType(section, type);
    return Object.entries(types).length;
}
exports.countResources = countResources;
function addEmptyProperties(template) {
    var _a;
    let section = (_a = template.Resources) !== null && _a !== void 0 ? _a : {};
    Object.keys(section).map((key) => {
        if (!section[key].hasOwnProperty('Properties')) {
            section[key].Properties = {};
        }
    });
    return template;
}
function filterType(section, type) {
    return Object.entries(section !== null && section !== void 0 ? section : {})
        .filter(([_, v]) => v.Type === type)
        .reduce((agg, [k, v]) => { return { ...agg, [k]: v }; }, {});
}
